File: Check Preorder of BST.cpp
// Problem: Check Preorder of BST
// geeksforgeeks problem of the day -> 23rd July 2026
// C++ CODE
class Solution {
  public:
    bool canRepresentBST(vector<int> &arr) {
        stack<int> st;
        int lowerBound = INT_MIN;
        for (int value : arr) {
            if (value < lowerBound)
                return false;
            while (!st.empty() && value > st.top()) {
                lowerBound = st.top();
                st.pop();
            }
            st.push(value);
        }
        return true;
    }
};

File: Check Preorder of BST.java
// Problem: Check Preorder of BST
// geeksforgeeks problem of the day -> 23rd July 2026
// JAVA CODE
class Solution {
    int index;
    
    public boolean canRepresentBST(List<Integer> arr) {
        index = 0;
        build(arr, Integer.MIN_VALUE, Integer.MAX_VALUE);
        return index == arr.size();
    }
    
    private void build(List<Integer> arr, int min, int max){
        if(index == arr.size()){
            return;
        }
        int val = arr.get(index);
        if(val < min || val > max){
            return;
        }
        
        index++;
        
        build(arr, min, val);
        build(arr, val, max);
    }
}